### PR TITLE
lib: misc: mtimer: Add MachineTimer

### PR DESCRIPTION
**Pull request by dnltz**
Some points I would like to have feedback to:
* I removed the 'Ctrl' from the bus components and only added it to the controller itself. I'm not sure if you need this suffix for bus components in e.g. SoC-files. I also wrote-out the shortcut for Ctrl...
* Added a 'connection' case class:
`case class MachineTimerControllerConnection(parameter: MachineTimerParameter) extends Bundle {
    val counter = out UInt(64 bits)
    val clear = in Bool
}
...
val io = new Bundle {
    val storage = in(MachineTimerControllerStorage(parameter))
    val connection = MachineTimerControllerConnection(parameter)
    val interrupt = out Bool
  }
`

### CODE REVIEW

**Comment by Dolu1990** [comment]
hmm for #190 and #189, maybe we should organize a chat on gitter, would that be ok ?

**Comment by Dolu1990** [comment]
It would be easier that way instead of having the git issues lattency ^^

**Comment by dnltz** [comment]
@Dolu1990 Worked a little bit on this. I also wrapped the bus stuff into an object. Unfortunately, using MachineTimer as extend does not work because slave(bus) failed due NullPointer. The bus variable is not set correctly at compile-time?!

File: lib/src/main/scala/spinal/lib/misc/mtimer/MachineTimer.scala [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+object MachineTimer {
+  case class Core(
+    parameter: MachineTimerCtrl.Parameter,
+    factory: BusSlaveFactory
+  ) extends Area {
+    val timerCtrl = MachineTimerCtrl(parameter)
+    val mapper = MachineTimerCtrl.Mapper(timerCtrl.io, factory, parameter)
+  }
+
+  case class Io[T <: IMasterSlave](busComp: T) extends Bundle {
```
Comment (Dolu1990) [inline lib/src/main/scala/spinal/lib/misc/mtimer/MachineTimer.scala]: T <: Data with IMasterSlave

busComp: HardType[T]   ;)

File: lib/src/main/scala/spinal/lib/misc/mtimer/MachineTimerCtrl.scala [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+
+    /* Start with initiale value of 1 to avoid a hit at reset */
+    val counter = Reg(UInt(p.width bits)) init(1)
+    io.counter := counter
+    val hit = RegInit(False)
+
+    counter := counter + 1
+    when (io.clear) {
+      hit := False
+    } .elsewhen (counter === io.config.compare) {
```
[the remark this answers is not in the document]
Comment (dnltz) [reply lib/src/main/scala/spinal/lib/misc/mtimer/MachineTimerCtrl.scala]: jap, that's right. Had no time to test this...

**Comment by Dolu1990** [comment]
I improved the initial design there : https://github.com/SpinalHDL/SaxonSoc/blob/Bmb/hardware/scala/saxon/MachineTimer.scala#L8

Seem to be just fine.
I can merge you pull request and patch this in on the top, or you wanna do it ? :)

Also, maybe we can just merge the Mapper class into the MachineTimerCtrl. actualy it seem a bit overenginnered, isn't it ?

**Comment by dnltz** [comment]
oh yeah the RegNext is nice and makes the code clean! I can merge them ... no problem for me.

Mhh I'll try some stuff with the Mapper class. There is indeed no need for the controller since the MachineTimes relies on a BusFactory ...

**Comment by dnltz** [comment]
So, I misunderstood your previous comment! I merged the Mapper into the controller.

One note about the init value of the compare register:
`val compare = Reg(UInt(p.width bits)) init(BigInt("FFFFFFFFFF", 16))`
Without this init value, it contains a random value and this could be a very low value, which can trigger a timer interrupt without reason before the scheduler is running. Therefore, this register has the maximum value of minimal supported reg width.

**Comment by Dolu1990** [comment]
About init(BigInt("FFFFFFFFFF", 16))
I'm not sure it is required. Is there anything about it in the spec ?
Normaly, you would not listen timer interrupt, until you configure it.

**Comment by dnltz** [comment]
"The interrupt will only be taken if interrupts are enabled and the MTIE bit is set in the mie register."
So I can remove it ^^

**Pull request closed without merging**